File: mkulima_app/lib/models/farm.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'farm.freezed.dart';
part 'farm.g.dart';

@freezed
class Farm with _$Farm {
  const factory Farm({
    required String uuid,
    required String name,
    required String location,
    @JsonKey(name: 'size_acres') required double sizeAcres,
    @JsonKey(name: 'crop_type') required String cropType,
    @JsonKey(name: 'soil_type') String? soilType,
    @JsonKey(name: 'planting_date') String? plantingDate,
    @JsonKey(name: 'harvest_expected_date') String? harvestExpectedDate,
    @Default('active') String status,
    String? notes,
    @Default([]) List<FarmActivity> activities,
  }) = _Farm;

  factory Farm.fromJson(Map<String, dynamic> json) => _$FarmFromJson(json);
}

@freezed
class FarmActivity with _$FarmActivity {
  const factory FarmActivity({
    required String uuid,
    @JsonKey(name: 'activity_type') required String activityType,
    @JsonKey(name: 'activity_date') required String activityDate,
    @JsonKey(name: 'cost_tzs') @Default(0.0) double costTzs,
    String? notes,
  }) = _FarmActivity;

  factory FarmActivity.fromJson(Map<String, dynamic> json) => _$FarmActivityFromJson(json);
}

File: app/Models/Farm.php
<?php

namespace App\Models;

use App\Traits\BelongsToTenant;
use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;
use Illuminate\Support\Str;

class Farm extends Model
{
    use BelongsToTenant, HasFactory;

    protected $fillable = [
        'uuid',
        'tenant_id',
        'user_id',
        'name',
        'location',
        'size_acres',
        'crop_type',
        'soil_type',
        'planting_date',
        'harvest_expected_date',
        'status',
        'notes',
    ];

    protected $casts = [
        'size_acres' => 'float',
        'planting_date' => 'date',
        'harvest_expected_date' => 'date',
    ];

    protected static function boot()
    {
        parent::boot();
        static::creating(function ($model) {
            if (empty($model->uuid)) {
                $model->uuid = (string) Str::uuid();
            }
        });
    }

    public function user(): BelongsTo
    {
        return $this->belongsTo(User::class);
    }

    public function activities(): HasMany
    {
        return $this->hasMany(FarmActivity::class);
    }
}
